src/bloqade/squin/analysis/shape/impls.py

```diff
@@ -0,0 +1,68 @@
+from typing import cast
+
+from kirin import ir, interp
+
+from bloqade.squin import op
+
+from .lattice import (
+    NoShape,
+    OpShape,
+)
+from .analysis import ShapeAnalysis
+
+
+@op.dialect.register(key="op.shape")
+class SquinOp(interp.MethodTable):
+
+    @interp.impl(op.stmts.Kron)
+    def kron(self, interp: ShapeAnalysis, frame: interp.Frame, stmt: op.stmts.Kron):
+        lhs = frame.get(stmt.lhs)
+        rhs = frame.get(stmt.rhs)
+        if isinstance(lhs, OpShape) and isinstance(rhs, OpShape):
+            new_size = lhs.size + rhs.size
+            return (OpShape(size=new_size),)
+        else:
+            return (NoShape(),)
+
+    @interp.impl(op.stmts.Mult)
+    def mult(self, interp: ShapeAnalysis, frame: interp.Frame, stmt: op.stmts.Mult):
+        lhs = frame.get(stmt.lhs)
+        rhs = frame.get(stmt.rhs)
+
+        if isinstance(lhs, OpShape) and isinstance(rhs, OpShape):
+            lhs_size = lhs.size
+            rhs_size = rhs.size
+            # Sized trait implicitly enforces that
+            # all operators are square matrices,
+            # not sure if it's worth raising an exception here
+            # or just letting this propagate...
+            if lhs_size != rhs_size:
```

Comment:
I feel like this should be an exception instead of just propagating a `NoShape?` because if you allow `Mult` to handle things that are not the same size you can get something that's not really valid or representable with the existing `Sized` trait